Add json, Uuid cast operator support

Sagebati opened the pull request:
I added JsonB and Uuid cast from text and to text support for the newish cast operator

I don't know, if cast() must compile check types that can be cast without failure (ex: int to text). If it's the case I can add a assume_cast() or faillible_cast() where the cast can fail.

guissalustiano commented on diesel/src/expression/cast.rs:
```diff
@@ -121,11 +122,15 @@ type_name! {
         Int4 => "integer",
         Int8 => "integer",
         Text => "char",
+
```
Could you remove this extra space?

Sagebati replied:
Good catch, done !

guissalustiano reviewed (approved):
Thanks for opening the PR! It looks good to me, just a minor extra space

Ten0 commented on diesel/src/expression/cast.rs:
```diff
@@ -96,10 +96,11 @@ where
 
 macro_rules! type_name {
-    ($($backend: ty: $backend_feature: literal { $($type: ident => $val: literal,)+ })*) => {
+    ($($backend: ty: $backend_feature: literal { $($($feature: literal;)? $type: ident => $val: literal,)+ })*) => {
```
It looks like `?` would feel more intuitive to read than `;` if possible.

Ten0 replied:
Closing because will be removed as per other comment. (I checked and indeed Uuid exists without the feature.)

Ten0 commented on diesel/src/expression/cast.rs:
```diff
@@ -121,11 +122,15 @@ type_name! {
         Interval => "interval",
         Time => "time",
         Timestamp => "timestamp",
+        "uuid"; Uuid => "uuid",
```
Is the Uuid **sql** type not defined unconditionally on whether the uuid feature is enabled? (If yes then cfg directives below should probably also be removed)

Sagebati replied:
I must misread the doc, thanks, I dropped the uuid feature

Ten0 commented on diesel/src/expression/cast.rs:
```diff
@@ -157,6 +162,16 @@ impl<ST1, ST2> CastsTo<sql_types::Nullable<ST2>> for sql_types::Nullable<ST1> wh
 {}
 
 impl CastsTo<sql_types::Int8> for sql_types::Int4 {}
+impl CastsTo<sql_types::Int8> for sql_types::Text {}
```
Yes I think I like the name `faillible_cast` you proposed for this kind of conversions: it feels like that would enable using the regular `cast` with more confidence, whereas otherwise every usage would have to be analyzed with great care when reviewing for correctness 🙂

Sagebati replied:
I personally prefer assume because the similarity with `assume_not_null()` but faillible is ok

Ten0 replied:
IMO clarity generally wins over similarity.

For `assume_not_null` I'm specifically trying to not convey information about behavior, because behavior may range from errors to wrong values being propagated into applicative code.

In this case however my understanding is that the only thing that can happen is an error at the database level, which is rather "safe" (compared to, say, incorrect values).

If I were to not know what this function does, and I were to read `assume_cast`, first I may not know whether that means that we assume that this is already casted, but even if I interpreted that correctly I wouldn't know what happens if the assumption is wrong, which would likely be something I'd have to look up in the documentation to assess the risk level and amount of care I have to put in making sure this never ever happens.

Generally for me "assume" is a keyword for "pay very special attention to what's written here, because the consequences if it's wrong are undefined", whereas here the consequences are well defined and it may useful for the user to know at a glance what they are.

While similarity may help with finding the function, I believe that error message customization on the trait bound that's required to access the infallible `cast()` version will cover well for the typical user story. That being said I'm not generally opposed to other names (e.g. if you had a strong feeling in favor of `try_cast()`). But `assume` is too intense for this case IMO.

Sagebati commented:
Took into account the feedback !

- Remove bad formatting 
- Removed the uuid feature
- Added more types to `CastTo
- Added `faillible_cast()` to support, user checked cast
- Added jsonb and json support for sqlite

Ten0 commented on diesel/src/expression_methods/global_expression_methods.rs:
```diff
@@ -483,6 +483,40 @@ pub trait ExpressionMethods: Expression + Sized {
         cast::Cast::new(self)
     }
 
+    /// Generates a `CAST(expr AS sql_type)` expression, this version does not check the castability, like [`cast()`](Self::cast).
```
Ooh I thought we were going to still check "castability" in the sense that it "may work depending on the value", but not "castability" in the sense that it "will always work from this type to that type".

Sagebati replied:
Maybe I misunderstood the initial concept, there are 3 concepts here:

1. I can always cast from this type T1 to T2 for all values.  => cast()
2. I can always cast from T1 to T2 for some values but some values may fail 
3. Free casting, only if the type is has a type_name => fallible_cast() 


Do you want me go for 2 instead of 3 ? I think it could be better yes

Ten0 replied:
Alright if we agree then let's go for version 2 😊

Sagebati replied:
Great ! 

Can Int8 -> Int4 fail ? Or is it just semantically wrong ?

Sagebati commented on diesel/src/expression/cast.rs:
```diff
@@ -149,14 +158,89 @@ where
+    ) => {
+        $(
+            $(#[cfg(feature = $feature)])?
+            impl FaillibleCastsTo<sql_types::$to> for sql_types::$from {}
```
I understand,the idea is that if when writing generic code, it's great to only depend on fallible_cast a la TryFrom and From. I can revert

Ten0 replied:
Ok then I'd tend to remove the requirement and supertrait constraint, and we can always add back the missing implementations in a backwards-compatible manner if people ask for them?

weiznich replied:
I personally think it's reasonable to have it in the current way to basically mirror what's `From` and `TryFrom` already do for rust types.

Ten0 replied:
Ok

Sagebati commented on diesel/src/expression/cast.rs:
```diff
@@ -149,14 +158,89 @@ where
+    (Int4 <- Float8),
+    (Int4 <- Text),
+    (Int8 <- Text),
+    (Float4 <- Int8),
```
Good catch

czotomo commented on diesel/src/expression/cast.rs:
```diff
@@ -149,14 +158,89 @@ where
 }
 
+/// Marker trait: this SQL type (`Self`) can be casted to the target SQL type, but some values can be invalid
+pub trait FaillibleCastsTo<ST> {}
```
Shouldn't it be `fallible`?

Sagebati commented:
Renamed Faillible to fallible, I think we are good, aren't we ?
